eventsources/sources/bitbucket/start.go

```diff
@@ -0,0 +1,344 @@
+	opt := &bitbucketv2.WebhooksOptions{
+		Owner:       es.Owner,
+		RepoSlug:    es.RepositorySlug,
+		Url:         formattedWebhookURL,
+		Description: "webhook managed by Argo-Events",
+		Active:      true,
+		Events:      es.Events,
+	}
+
+	return router.client.Repositories.Webhooks.Create(opt)
+}
+
+// updateWebhook updates an existing webhook
+func (router *Router) updateWebhook(existingHookSubscription *WebhookSubscription) (*bitbucketv2.Webhook, error) {
+	es := router.bitbucketEventSource
+	opt := &bitbucketv2.WebhooksOptions{
+		Owner:       es.Owner,
+		RepoSlug:    es.RepositorySlug,
+		Uuid:        existingHookSubscription.Uuid,
+		Description: existingHookSubscription.Description,
+		Url:         existingHookSubscription.Url,
+		Active:      existingHookSubscription.Active,
+		Events:      es.Events,
+	}
+
+	return router.client.Repositories.Webhooks.Update(opt)
+}
+
+// deleteWebhook deletes an existing webhook
+func (router *Router) deleteWebhook(hookID string) error {
+	es := router.bitbucketEventSource
+	_, err := router.client.Repositories.Webhooks.Delete(&bitbucketv2.WebhooksOptions{
+		Owner:    es.Owner,
+		RepoSlug: es.RepositorySlug,
+		Uuid:     hookID,
+	})
+
+	return err
+}
+
+// listWebhooks gets a list of all existing webhooks in target repository
+func (router *Router) listWebhooks() ([]WebhookSubscription, error) {
+	es := router.bitbucketEventSource
+	hooksResponse, err := router.client.Repositories.Webhooks.Gets(&bitbucketv2.WebhooksOptions{
+		Owner:    es.Owner,
+		RepoSlug: es.RepositorySlug,
+		Active:   true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return router.extractHooksFromListResponse(hooksResponse)
+}
+
+// extractHooksFromListResponse helper that extracts the list of webhooks from the response of listWebhooks
+func (router *Router) extractHooksFromListResponse(listHooksResponse interface{}) ([]WebhookSubscription, error) {
+	logger := router.GetRoute().Logger
+	res, ok := listHooksResponse.(map[string]interface{})
+	if !ok {
+		logger.Errorw("failed to parse the list webhooks response", zap.Any("response", listHooksResponse))
+		return nil, fmt.Errorf("failed to parse the list webhooks response")
+	}
+
+	var hooks []WebhookSubscription
+	err := mapstructure.Decode(res["values"], &hooks)
+	if err != nil || hooks == nil {
+		logger.Errorw("failed to parse the list webhooks response", zap.Any("response", listHooksResponse))
+		return nil, fmt.Errorf("failed to parse the list webhooks response")
+	}
+
+	return hooks, nil
+}
+
+// findWebhook searches for a webhook in a list by its URL and returns the webhook if its found
+func (router *Router) findWebhook(hooks []WebhookSubscription, targetWebhookURL string) (*WebhookSubscription, bool) {
+	var existingHookSubscription *WebhookSubscription
+	isFound := false
+	for _, hook := range hooks {
+		if hook.Url == targetWebhookURL {
+			isFound = true
+			existingHookSubscription = &hook
+			router.hookID = hook.Uuid
+			break
```

Comment:
Instead of `break` and defining variables `existingHookSubscription` and `isFound`, we can just return &WebhookSubscription{}, true on 332 and nil, false on 336.

Also, what's the purpose of `isFound`? can the caller not just check whether the pointer to WebhookSubscription is nil or not?

Reply:
Yes of course you could replace the break, I just wanted to avoid an extra `return` statement if it's unnecessary, and thought it would be more readable this way. Same regarding `isFound` - just thought it would be more readable than comparing the pointer to nil. Just personal taste I guess